FSE: Add basic template information to editor header

noahtallen opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/master/docs/contributors/repository-management.md#pull-requests. -->

## Description
Adds a basic "document actions" component which is rendered in the center of the FSE editor header. Note: the string rendered in the title is TBD. Currently it uses the template name, but we are also thinking that the page name could be more useful / easier to understand.

Also, when a template part block is directly selected, it will show its label as secondary text in the header area. We are hoping to expand this functionality to include hover, and to handle if the parent of the selected block is the template part.

This is also meant to serve as a starting point for adding a document settings dropdown to the header.

to do :
- [x] Finalize the component API (what the props should be called and how they should be passed)
- [x] Should we include the temporary dropdown now or merge without it?
- [x] Figure out how to store the "secondary" item and where to pull that data from.

Work towards #25085. cc @dubielzyk 

## How has this been tested?
locally in edit site

## Screenshots <!-- if applicable -->

_A friendly reminder that eventually, the "index" text in the left will go away._
![2020-09-17 15 27 12](https://user-images.githubusercontent.com/6265975/93535171-4b2de480-f8fb-11ea-81b9-9c0dc665629e.gif)


## Types of changes
Enhancement

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

noahtallen committed:
Add basic component centered in edit site header

noahtallen committed:
Add classnames dependency to edit-site

noahtallen committed:
Add basic dropdown component to header item

noahtallen commented on packages/edit-site/src/components/header/index.js:
```diff
@@ -81,62 +82,73 @@ export default function Header( {
-					label={ _x(
-						'Add block',
-						'Generic label for block inserter button'
+			<div className="edit-site-header_start">
```
The changes in this file add three wrapper divs around the content of the header. One for the "start", "center", and "end". The start and end divs have `flex: 1`. This way, the center div is always centered in the middle of the viewport regardless of the width of the content inside the start/end divs.

noahtallen commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -0,0 +1,71 @@
+							<Button
+								onClick={ onToggle }
+								className="edit-site-document-actions__document_root"
+								aria-haspopup="true"
```
I'm unsure which props we need to add to the button

Addison-Stavlo replied:
Im not certain about a lot of the standard ones, but this seems like a good start.  However, it may also be useful to add a function `onKeyDown` if we want to stay ahead on accessibility:
https://github.com/WordPress/gutenberg/blob/af768dc98c625e33de8023084c6a5c198c52f8ef/packages/block-library/src/heading/heading-level-dropdown.js#L49-L55

noahtallen replied:
Nice, thanks for that. Adjusted the props to match that and it's looking much better

noahtallen committed:
Add extra props to dropdown button

noahtallen committed:
Move template fetch to header component

shaunandrews commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -0,0 +1,71 @@
+	return (
+		<div className="edit-site-document-actions">
+			{ primaryText ? (
+				<Dropdown
```
I'm not sure where we've landed on this being a dropdown; I think it makes a lot of sense for the document/template label to open some settings, but I'm not so sure about the template part label. In some designs the template part label could be useful as a way to select the template part itself. This could confuse the interactions between dropdown and selection.

For now, I'd suggest we move forward with _just_ the labels as indicators of what template and template part of being edited.

Addison-Stavlo replied:
> In some designs the template part label could be useful as a way to select the template part itself.

If the label already shows the name of the template part, is it not already selected or hovered?  And in the case of being hovered you couldn't hover it and click the name in the top-bar at the same time so... I guess I dont understand how clicking the name of the Template Part here would act to select it?

Addison-Stavlo replied:
> I'm not sure where we've landed on this being a dropdown; I think it makes a lot of sense for the document/template label to open some settings

So what should it open for controlling the settings then?  I thought the current design mocks for these settings were a dropdown?

shaunandrews replied:
It might be best to focus on showing the template part being edited first — and then add the interaction in another PR. I'll defer to @dubielzyk here.

noahtallen replied:
that sounds fine to me. I used the dropdown because I thought the document settings menu would be activated on click of this "area," and based on current designs that does at least sort of look like a dropdown menu 😅 

Totally on board with keeping just the labels for the time being.

dubielzyk replied:
>In some designs the template part label could be useful as a way to select the template part itself. This could confuse the interactions between dropdown and selection.

I think we should use the same treatment for both. The behavior should be similar between template and page/post. If we agree that there will be settings in there at some point, I'd say we should use the dropdown interaction.

If we use a staggered approach I think we could push it out in phases:
Phase 1: Show label. Top label is template or page/post. Bottom label is template part.
Phase 2: Add dropdown interaction. Add renaming of Template or Page/Post name. [Mocks](
![image](https://user-images.githubusercontent.com/2896062/93277004-73b1c380-f7c1-11ea-9b2b-dfff58d17a74.png))
Phase 3: Start looking at adding relevant settings

How does that sound to everyone? @shaunandrews @noahtallen @Addison-Stavlo

noahtallen replied:
yep! Per:

> Top label is template or page/post

I think this opens a can of worms conceptually, because if we go with "page," we're starting to think about the site editor from a user's page-first perspective rather than a technical person's template-first perspective. To be clear, it's easy to change the variable rendering this technically :D


I'm mostly convinced by the argument to think about the site editor from a page-first perspective.

dubielzyk replied:
I 100% agree that we should go page/post first. I think we only show the template if they use an entry-point that is about editing a specific template.

noahtallen replied:
>  I think we only show the template 

right, and to be clear, the template would get "shown" in the editor either way, but it would be more of the focus there

dubielzyk replied:
I mean this:
![image](https://user-images.githubusercontent.com/2896062/93285443-20497080-f7d5-11ea-88a4-fff45876491f.png)

So yeah, we'd show a way to see what template is being affected either way, but if a users wants to edit their _About_ page, then we'd display About in the top bar. If they are going to editing a specific template, we'd show that in the top bar. Hope that makes more sense 😄

noahtallen replied:
That does make sense. I think, as we discussed in the meeting, this is tricky because there are no entrypoints currently which let us infer if the user wants to mainly edit a template or a page. :)

jeyip reviewed (commented):
Nice job! These changes are compact, straightforward, and I think they'll serve as a great point of discussion for design.

As far as the code goes, I don't see any red flags or areas of concern. It also seems like the work here can be adapted very easily to the new mocks proposed by Shaun. 👍

noahtallen committed:
Use vertical design for header information

noahtallen commented:
cc @shaunandrews updated with vertical design:

![2020-09-15 11 50 23](https://user-images.githubusercontent.com/6265975/93252130-ed0dcf80-f749-11ea-8cbe-14ef2816c1ec.gif)

noahtallen commented on packages/edit-site/src/components/header/document-actions/style.scss:
```diff
@@ -0,0 +1,14 @@
+	flex-direction: column;
+	align-items: center;
+	justify-content: space-evenly;
+	margin: 5px 0;
```
I'm not a huge fan of how this margin is different from the rest of the items in the header, but it does match the latest in #25085 🤔

noahtallen commented:
Trying out the template part selection (doesn't handle parent blocks yet)

![2020-09-15 12 18 48](https://user-images.githubusercontent.com/6265975/93254756-b46ff500-f74d-11ea-96dd-8c8395dfe01b.gif)

noahtallen committed:
Add hook, rework props

noahtallen committed:
Update spacing

noahtallen committed:
Add CSS transition

noahtallen commented:
In 202d122, I added the animation from the design, and also corrected the spacing to match [this comment](https://github.com/WordPress/gutenberg/issues/25085#issuecomment-693008007).

Addison-Stavlo commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -0,0 +1,80 @@
+		return select( 'core/block-editor' ).getSelectedBlock();
+	} );
+
+	// TODO: Handle if parent is template part too.
```
I wonder how to go about this here.  I was looking previously and the `getBlockParents` selector didn't seem to return the template part block as a parent for its children.  That may be a complication of of the template part block referencing a separate entity? 🤔 

One idea might be to neglect looking at the selected block in this component and instead dispatching from the template part block itself since it can easily determine `isSelected || hasInnerBlockSelected`.

noahtallen replied:
Maybe we could implement a selector in the block-editor store like `getNearestInnerBlockController` or something like that which could avoid the parent state problem

Addison-Stavlo replied:
Yeah possibly...  Although if there is an inner block controller that we aren't concerned with nested inside a template part, we may need to go beyond the nearest parent controller? 🤔   Adding something like that may be useful in other places as well, but if it is messy dispatching from the block might be very simplified.

noahtallen replied:
Hm, true. I guess that design would assume that any inner block controller could show up in the header (even if created by a 3rd party)

Addison-Stavlo commented on packages/edit-site/src/components/header/document-actions/index.js:
```diff
@@ -0,0 +1,80 @@
+}
+
+export default function DocumentActions( { documentTitle } ) {
+	const { label, isActive } = useSecondaryText();
```
why do we need `isActive` ?   `isActive` seems to always be true when there is a label, and undefined when the label is undefined.  

Do we need a fallback/placeholder name for when `selectedBlockLabel` is an empty string or something?

noahtallen replied:
I was thinking ahead to hover state, where if you were hovering the item, it might be grayed out until you select it. But that's TBD I guess, so it doesn't do a whole lot right now!

Addison-Stavlo reviewed (commented):
This is looking good and behaving well to me! 

As noted in `TODO` we don't have the child selected state covered, but I would be happy with this being addressed in a follow-up (potentially along with the hover label).

Addison-Stavlo reviewed (approved):
This works well and looks good to me!

I noticed the template title isn't capitalized for index/singular on my end.  Do we need to use a title/label instead of the slug?  (not to block this, however)

noahtallen commented:
> I noticed the template title isn't capitalized for index/singular on my end. Do we need to use a title/label instead of the slug? (not to block this, however)

nice catch! Currently the title is equal to the slug, so it also isn't capitalized. I'm personally inclined to leave it as is since we'll probably switch to use the page context soon. Plus, the index slug is probably the most accurate thing to use at this point since we don't have a solid system for having a "display name" for a template
